helix-term/src/application.rs

```diff
@@ -827,19 +901,56 @@ impl Application {
         }));
 
         self.event_loop(input_stream).await;
+
         self.close().await?;
         self.restore_term()?;
 
         Ok(self.editor.exit_code)
     }
 
     pub async fn close(&mut self) -> anyhow::Result<()> {
-        self.jobs.finish().await?;
-
-        if self.editor.close_language_servers(None).await.is_err() {
-            log::error!("Timed out waiting for language servers to shutdown");
+        // [NOTE] we intentionally do not return early for errors because we
+        //        want to try to run as much cleanup as we can, regardless of
+        //        errors along the way
+
+        let mut result = match self
+            .jobs
+            .finish(Some(&mut self.editor), Some(&mut self.compositor))
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                log::error!("Error executing job: {}", err);
+                Err(err)
+            }
         };
 
-        Ok(())
+        for doc in self.editor.documents_mut() {
+            if let Some(save_result) = doc.close().await {
+                result = match save_result {
+                    Ok(_) => result,
+                    Err(err) => {
+                        if let Some(path) = doc.path() {
+                            log::error!(
+                                "Error saving document '{}': {}",
+                                path.to_string_lossy(),
+                                err
+                            );
+                        }
+                        Err(err)
+                    }
+                };
+            }
+        }
+
+        match self.editor.close_language_servers(None).await {
```

Comment:
I wonder if we still need the argument given that this is not used in other places, might as well just drop it.

Reply:
Hmm, being able to configure an LSP shutdown timeout seems useful though, especially given we've had recent bug reports like #1639. I think we just need to do the work of adding a config option for some of the timeouts for LSP requests.